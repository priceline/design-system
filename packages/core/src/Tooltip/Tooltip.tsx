import { themeGet } from '@styled-system/theme-get'
import React from 'react'
import styled, { withTheme } from 'styled-components'
import { BorderRadiusProps } from 'styled-system'
import { Box } from '../Box/Box'
import { PaletteColor } from '../theme/theme'
import { applyVariations, colorScheme, getPaletteColor } from '../utils/utils'

const arrowShadow = (props) => {
  return props.top
    ? {
        'box-shadow': '-9.66px 9.66px 8px 0 rgba(0,0,0,0.04), -4px 4px 4px 0 rgba(0,0,0,0.08)',
      }
    : {
        'box-shadow': '-1.41px 1.41px 1px 0 rgba(0,0,0,0.01), -3.66px 3.66px 8px 0 rgba(0,0,0,0.04)',
      }
}

const arrowAlign = (props) => {
  return props.left
    ? { left: '16px', 'margin-left': props.top ? 0 : '15px' }
    : props.center
    ? { left: '50%', 'margin-left': props.top ? '-7px' : '7px' }
    : { right: '16px', 'margin-right': props.top ? '5px' : '-10px' }
}

const arrowPosition = (props) => {
  return props.top
    ? {
        'transform-origin': '0 0',
        transform: 'rotate(-45deg)',
        bottom: '-10px',
      }
    : {
        'transform-origin': '0 0',
        transform: 'rotate(-225deg)',
        top: '0',
      }
}

const arrow = (props) => {
  return props.top
    ? {
        'transform-origin': '0 0',
        transform: 'rotate(-45deg)',
      }
    : {
        'transform-origin': '0 0',
        transform: 'rotate(-225deg)',
      }
}

const tooltipPosition = (props) => {
  return props.top ? { bottom: '-8px' } : { top: 0 }
}

const tooltipAlign = (props) => {
  return props.right
    ? { right: 0 }
    : props.center
    ? { left: '50%', width: 'auto', transform: 'translateX(-50%)' }
    : null
}

const TooltipContent = styled(Box)`
  display: inline;
  font-size: ${({ theme }) => theme.fontSizes[0]}px;
  position: absolute;
  box-sizing: border-box;
  background: ${(props) => getPaletteColor(props.bg || props.color, 'base')(props)};
  text-align: center;

  ${tooltipPosition} ${tooltipAlign} &::after {
    content: '';
    position: absolute;
    width: 0;
    height: 0;
    border-width: 5px;
    border-style: solid;
    border-color: transparent transparent
      ${(props) => getPaletteColor(props.bg || props.color, 'base')(props)}
      ${(props) => getPaletteColor(props.bg || props.color, 'base')(props)};

    ${arrow} ${arrowPosition} ${arrowAlign} ${arrowShadow};
    ${(props) => {
      const colorScheme = themeGet(`colorSchemes.${props.colorScheme}`)(props)
      return colorScheme
        ? `
        color: ${colorScheme.background};
        border-color: ${colorScheme.background};
      `
        : ''
    }}
  }

  ${colorScheme}

  ${applyVariations('Tooltip')}
`

/**
 * @public
 */
export type TooltipProps = BorderRadiusProps & {
  children?: React.ReactNode
  bg?: PaletteColor
  bottom?: boolean
  center?: boolean
  color?: PaletteColor
  left?: boolean
  position?: string
  right?: boolean
  top?: boolean
  zIndex?: number | string
}

function _Tooltip({
  children,
  borderRadius = 'md',
  bottom = true,
  color = 'background.lightest',
  zIndex = 'auto',
  ...props
}: TooltipProps): React.ReactElement {
  const defaultedProps = {
    borderRadius,
    color,
    zIndex,
    ...props,
  }
  return (
    <div style={{ position: 'relative', zIndex: defaultedProps.zIndex }}>
      <TooltipContent p={2} mb={3} mt={2} boxShadowSize='md' {...defaultedProps}>
        {children}
      </TooltipContent>
    </div>
  )
}

/**
 * @public
 */
export const Tooltip = withTheme(_Tooltip)
