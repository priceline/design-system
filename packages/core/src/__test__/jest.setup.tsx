import '@testing-library/jest-dom'
import 'jest-styled-components'

global.requestAnimationFrame =
  global.requestAnimationFrame ||
  function _raf(cb) {
    return setTimeout(cb, 0)
  }
