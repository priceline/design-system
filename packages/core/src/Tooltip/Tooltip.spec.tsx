import { render } from '../__test__/testing-library'
import { Tooltip } from './Tooltip'

describe('Tooltip', () => {
  test('renders', () => {
    const { asFragment } = render(<Tooltip>A Tooltip</Tooltip>)
    expect(asFragment()).toMatchSnapshot()
  })

  test('positioning variations', () => {
    const { rerender, asFragment } = render(
      <Tooltip color='primary' top left>
        left tooltip
      </Tooltip>
    )
    expect(asFragment()).toMatchSnapshot()

    rerender(
      <Tooltip color='primary' top center>
        centered tooltip
      </Tooltip>
    )
    expect(asFragment()).toMatchSnapshot()

    rerender(
      <Tooltip color='error' top right>
        right tooltip
      </Tooltip>
    )
    expect(asFragment()).toMatchSnapshot()

    rerender(
      <Tooltip bottom left>
        left tooltip
      </Tooltip>
    )
    expect(asFragment()).toMatchSnapshot()

    rerender(
      <Tooltip bottom center>
        centered tooltip
      </Tooltip>
    )
    expect(asFragment()).toMatchSnapshot()

    rerender(
      <Tooltip bottom right>
        right tooltip
      </Tooltip>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
