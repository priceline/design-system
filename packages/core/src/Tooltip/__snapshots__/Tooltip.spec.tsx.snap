// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip positioning variations 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-color: #0068ef;
  color: #fff;
  padding: 8px;
  margin-bottom: 16px;
  margin-top: 8px;
  border-radius: 8px;
  box-shadow: 0 -1px 0 0 rgba(0,0,0,0.03),0 0 2px 0 rgba(0,0,0,0.2),0 4px 2px -2px rgba(0,0,0,0.12),0 4px 8px -1px rgba(0,0,0,0.16);
}

.c2 {
  display: inline;
  font-size: 12px;
  position: absolute;
  box-sizing: border-box;
  background: #0068ef;
  text-align: center;
  bottom: -8px;
}

.c2::after {
  content: '';
  position: absolute;
  width: 0;
  height: 0;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent #0068ef #0068ef;
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  bottom: -10px;
  left: 16px;
  margin-left: 0;
  box-shadow: -9.66px 9.66px 8px 0 rgba(0,0,0,0.04),-4px 4px 4px 0 rgba(0,0,0,0.08);
}

<div
    class="c0"
  >
    <div
      style="position: relative; z-index: auto;"
    >
      <div
        class="c1 c2"
        color="primary"
      >
        left tooltip
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip positioning variations 2`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-color: #0068ef;
  color: #fff;
  padding: 8px;
  margin-bottom: 16px;
  margin-top: 8px;
  border-radius: 8px;
  box-shadow: 0 -1px 0 0 rgba(0,0,0,0.03),0 0 2px 0 rgba(0,0,0,0.2),0 4px 2px -2px rgba(0,0,0,0.12),0 4px 8px -1px rgba(0,0,0,0.16);
}

.c2 {
  display: inline;
  font-size: 12px;
  position: absolute;
  box-sizing: border-box;
  background: #0068ef;
  text-align: center;
  bottom: -8px;
  left: 50%;
  width: auto;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c2::after {
  content: '';
  position: absolute;
  width: 0;
  height: 0;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent #0068ef #0068ef;
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  bottom: -10px;
  left: 50%;
  margin-left: -7px;
  box-shadow: -9.66px 9.66px 8px 0 rgba(0,0,0,0.04),-4px 4px 4px 0 rgba(0,0,0,0.08);
}

<div
    class="c0"
  >
    <div
      style="position: relative; z-index: auto;"
    >
      <div
        class="c1 c2"
        color="primary"
      >
        centered tooltip
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip positioning variations 3`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-color: #c00;
  color: #fff;
  padding: 8px;
  margin-bottom: 16px;
  margin-top: 8px;
  border-radius: 8px;
  box-shadow: 0 -1px 0 0 rgba(0,0,0,0.03),0 0 2px 0 rgba(0,0,0,0.2),0 4px 2px -2px rgba(0,0,0,0.12),0 4px 8px -1px rgba(0,0,0,0.16);
}

.c2 {
  display: inline;
  font-size: 12px;
  position: absolute;
  box-sizing: border-box;
  background: #c00;
  text-align: center;
  bottom: -8px;
  right: 0;
}

.c2::after {
  content: '';
  position: absolute;
  width: 0;
  height: 0;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent #c00 #c00;
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  bottom: -10px;
  right: 16px;
  margin-right: 5px;
  box-shadow: -9.66px 9.66px 8px 0 rgba(0,0,0,0.04),-4px 4px 4px 0 rgba(0,0,0,0.08);
}

<div
    class="c0"
  >
    <div
      style="position: relative; z-index: auto;"
    >
      <div
        class="c1 c2"
        color="error"
      >
        right tooltip
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip positioning variations 4`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-color: #fff;
  color: #001833;
  padding: 8px;
  margin-bottom: 16px;
  margin-top: 8px;
  border-radius: 8px;
  box-shadow: 0 -1px 0 0 rgba(0,0,0,0.03),0 0 2px 0 rgba(0,0,0,0.2),0 4px 2px -2px rgba(0,0,0,0.12),0 4px 8px -1px rgba(0,0,0,0.16);
}

.c2 {
  display: inline;
  font-size: 12px;
  position: absolute;
  box-sizing: border-box;
  background: #fff;
  text-align: center;
  top: 0;
}

.c2::after {
  content: '';
  position: absolute;
  width: 0;
  height: 0;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent #fff #fff;
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-225deg);
  -ms-transform: rotate(-225deg);
  transform: rotate(-225deg);
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-225deg);
  -ms-transform: rotate(-225deg);
  transform: rotate(-225deg);
  top: 0;
  left: 16px;
  margin-left: 15px;
  box-shadow: -1.41px 1.41px 1px 0 rgba(0,0,0,0.01),-3.66px 3.66px 8px 0 rgba(0,0,0,0.04);
}

<div
    class="c0"
  >
    <div
      style="position: relative; z-index: auto;"
    >
      <div
        class="c1 c2"
        color="background.lightest"
      >
        left tooltip
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip positioning variations 5`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-color: #fff;
  color: #001833;
  padding: 8px;
  margin-bottom: 16px;
  margin-top: 8px;
  border-radius: 8px;
  box-shadow: 0 -1px 0 0 rgba(0,0,0,0.03),0 0 2px 0 rgba(0,0,0,0.2),0 4px 2px -2px rgba(0,0,0,0.12),0 4px 8px -1px rgba(0,0,0,0.16);
}

.c2 {
  display: inline;
  font-size: 12px;
  position: absolute;
  box-sizing: border-box;
  background: #fff;
  text-align: center;
  top: 0;
  left: 50%;
  width: auto;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c2::after {
  content: '';
  position: absolute;
  width: 0;
  height: 0;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent #fff #fff;
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-225deg);
  -ms-transform: rotate(-225deg);
  transform: rotate(-225deg);
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-225deg);
  -ms-transform: rotate(-225deg);
  transform: rotate(-225deg);
  top: 0;
  left: 50%;
  margin-left: 7px;
  box-shadow: -1.41px 1.41px 1px 0 rgba(0,0,0,0.01),-3.66px 3.66px 8px 0 rgba(0,0,0,0.04);
}

<div
    class="c0"
  >
    <div
      style="position: relative; z-index: auto;"
    >
      <div
        class="c1 c2"
        color="background.lightest"
      >
        centered tooltip
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip positioning variations 6`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-color: #fff;
  color: #001833;
  padding: 8px;
  margin-bottom: 16px;
  margin-top: 8px;
  border-radius: 8px;
  box-shadow: 0 -1px 0 0 rgba(0,0,0,0.03),0 0 2px 0 rgba(0,0,0,0.2),0 4px 2px -2px rgba(0,0,0,0.12),0 4px 8px -1px rgba(0,0,0,0.16);
}

.c2 {
  display: inline;
  font-size: 12px;
  position: absolute;
  box-sizing: border-box;
  background: #fff;
  text-align: center;
  top: 0;
  right: 0;
}

.c2::after {
  content: '';
  position: absolute;
  width: 0;
  height: 0;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent #fff #fff;
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-225deg);
  -ms-transform: rotate(-225deg);
  transform: rotate(-225deg);
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-225deg);
  -ms-transform: rotate(-225deg);
  transform: rotate(-225deg);
  top: 0;
  right: 16px;
  margin-right: -10px;
  box-shadow: -1.41px 1.41px 1px 0 rgba(0,0,0,0.01),-3.66px 3.66px 8px 0 rgba(0,0,0,0.04);
}

<div
    class="c0"
  >
    <div
      style="position: relative; z-index: auto;"
    >
      <div
        class="c1 c2"
        color="background.lightest"
      >
        right tooltip
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip renders 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-color: #fff;
  color: #001833;
  padding: 8px;
  margin-bottom: 16px;
  margin-top: 8px;
  border-radius: 8px;
  box-shadow: 0 -1px 0 0 rgba(0,0,0,0.03),0 0 2px 0 rgba(0,0,0,0.2),0 4px 2px -2px rgba(0,0,0,0.12),0 4px 8px -1px rgba(0,0,0,0.16);
}

.c2 {
  display: inline;
  font-size: 12px;
  position: absolute;
  box-sizing: border-box;
  background: #fff;
  text-align: center;
  top: 0;
}

.c2::after {
  content: '';
  position: absolute;
  width: 0;
  height: 0;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent #fff #fff;
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-225deg);
  -ms-transform: rotate(-225deg);
  transform: rotate(-225deg);
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-225deg);
  -ms-transform: rotate(-225deg);
  transform: rotate(-225deg);
  top: 0;
  right: 16px;
  margin-right: -10px;
  box-shadow: -1.41px 1.41px 1px 0 rgba(0,0,0,0.01),-3.66px 3.66px 8px 0 rgba(0,0,0,0.04);
}

<div
    class="c0"
  >
    <div
      style="position: relative; z-index: auto;"
    >
      <div
        class="c1 c2"
        color="background.lightest"
      >
        A Tooltip
      </div>
    </div>
  </div>
</DocumentFragment>
`;
